### ISSUE TITLE
UnicodeDecodeError on python 2.x

### ISSUE THREAD

**Issue by ash30** [miguelgrinberg/python-socketio#239]
Hi,

I can successfully connect to a 3rd party socketio server when using python 3.x but when attempting to use the library on python 2.7.10 I'm getting a unicode decode error.

I assume its the 3rd party server sending unicode and my python socket client expecting ascii, is this a known issue for the 2.x library? 

thanks for your time

```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/usera-org/.local/share/virtualenvs/socket_test-nQGbRhBZ/lib/python2.7/site-packages/engineio/client.py", line 579, in _write_loop
    headers={'Content-Type': 'application/octet-stream'})
  File "/Users/usera-org/.local/share/virtualenvs/socket_test-nQGbRhBZ/lib/python2.7/site-packages/engineio/client.py", line 406, in _send_request
    return self.http.request(method, url, headers=headers, body=body)
  File "/Users/usera-org/.local/share/virtualenvs/socket_test-nQGbRhBZ/lib/python2.7/site-packages/urllib3/request.py", line 72, in request
    **urlopen_kw)
  File "/Users/usera-org/.local/share/virtualenvs/socket_test-nQGbRhBZ/lib/python2.7/site-packages/urllib3/request.py", line 150, in request_encode_body
    return self.urlopen(method, url, **extra_kw)
  File "/Users/usera-org/.local/share/virtualenvs/socket_test-nQGbRhBZ/lib/python2.7/site-packages/urllib3/poolmanager.py", line 323, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
  File "/Users/usera-org/.local/share/virtualenvs/socket_test-nQGbRhBZ/lib/python2.7/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/Users/usera-org/.local/share/virtualenvs/socket_test-nQGbRhBZ/lib/python2.7/site-packages/urllib3/connectionpool.py", line 354, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 1042, in request
    self._send_request(method, url, body, headers)
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 1082, in _send_request
    self.endheaders(body)
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 1038, in endheaders
    self._send_output(message_body)
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 880, in _send_output
    msg += message_body
UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 2: ordinal not in range(128)
```

**Comment by miguelgrinberg** [comment]
@ash30 this is when sending data from your application. Are you working with binary data by any chance?

**Comment by ash30** [comment]
It doesn't seem to be binary data. I delved into it today, from what I can see:

- `message_body = '\x00\x01\xff2'`
- `msg = u'POST /rtc/?transport=polling&EIO=3&sid=udFZMakBliNjq21mAAAM HTTP/1.1\r\nHost: localhost:8000\r\nAccept-Encoding: identity\r\nContent-Length: 4\r\nContent-Type: application/octet-stream\r\n\r\n'`
- Going up the frames, this message seems to be from an Array of x1 packet: `[{'data': None, 'packet_type': 2, 'binary': False}]`

I can reproduce using the following tests scripts:
Server ( js )
```
const socket_io = require("socket.io")({
    path: "/rtc"
  })
const restify = require("restify")

var defaultHandler = function(req, res) {
  res.send("hello")
};

const httpServer = restify.createServer({name:'Test Server'}, defaultHandler);
const io = socket_io.listen(httpServer.server);

io.on("connection", socket => {
    console.log("connected")
});

httpServer.listen(8000);
```

Client
```
import socketio
if __name__ == "__main__":
    url = sys.argv[1]
    socketio.Client().connect(url,socketio_path="/rtc")
```

Library Version
- Python 2.7.13
- python-engineio  3.2.3
- python-socketio  3.1.1

Haven't quite worked out whats happening, I'll try and take a look again when I can.

**Comment by miguelgrinberg** [comment]
I don't see any errors when running your node server and Python client.

**Comment by imcom** [comment]
Hi guys, I have the very same issue with @ash30 

Library Version
Python 2.7.15
python-socketio                    3.1.1
python-engineio                    3.3.0
python-engineio-client             1.1

just a connect will trigger the issue
```
sio.connect(self.chatroom_addr)
```

**Comment by imcom** [comment]
Use python3 solves the problem ...

**Comment by miguelgrinberg** [comment]
@imcom do you get an error in the client or the server? The OP was having issues with the client, but you are showing a server...

**Comment by saknarak** [comment]
I have same problem in python27, no problem in python3

my server is nodejs
nodejs and python3 client connect without problem
but python27 have a problem with encoding

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/engineio/client.py", line 578, in _write_loop
    headers={'Content-Type': 'application/octet-stream'})
  File "/usr/local/lib/python2.7/dist-packages/engineio/client.py", line 406, in _send_request
    return self.http.request(method, url, headers=headers, body=body)
  File "/usr/local/lib/python2.7/dist-packages/urllib3/request.py", line 72, in request
    **urlopen_kw)
  File "/usr/local/lib/python2.7/dist-packages/urllib3/request.py", line 150, in request_encode_body
    return self.urlopen(method, url, **extra_kw)
  File "/usr/local/lib/python2.7/dist-packages/urllib3/poolmanager.py", line 323, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
  File "/usr/local/lib/python2.7/dist-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python2.7/dist-packages/urllib3/connectionpool.py", line 354, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python2.7/httplib.py", line 1042, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1082, in _send_request
    self.endheaders(body)
  File "/usr/lib/python2.7/httplib.py", line 1038, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 880, in _send_output
    msg += message_body
UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 3: ordinal not in range(128)
```

I try to change encoding
```
import sys

reload(sys)
sys.setdefaultencoding('utf8')
```

but still not solve the problem, got this stack trace instead
```
...
  File "/usr/lib/python2.7/httplib.py", line 880, in _send_output
    msg += message_body
  File "/usr/lib/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xff in position 3: invalid start byte
```

**Comment by miguelgrinberg** [comment]
@saknarak same question I asked above, that nobody answered yet:

Are you working with binary data by any chance?

**Comment by saknarak** [comment]
Binary data is not from my json data but from socketio protocol (maybe)
from /usr/local/lib/python2.7/dist-packages/engineio/client.py line 576
I print p.encode() out
I found there are binary data between my data
![image](https://user-images.githubusercontent.com/7794793/51928191-e7099480-2427-11e9-8dd6-ef940fe5a761.png)

Now I can fix my problem by apply this patch

https://bugs.python.org/file21747/python-2.7.1-fix-httplib-UnicodeDecodeError.patch

BEFORE PATCH
```
         if isinstance(message_body, str):
            msg += message_body
            message_body = None
         self.send(msg)
```
AFTER PATCH
```
         if isinstance(message_body, str):
             try:
                 msg += message_body
                 message_body = None
             except UnicodeDecodeError:
                 pass
         self.send(msg)
```

after this patch, my code work perfectly

**Comment by saikrishb** [comment]
Got around this issue by converting the _sid_ (unicode) to a string in engineio/client.py (line 286)
The sid is later concatenated with the _base_url_ which is a string (line 291)

**Comment by imcom** [comment]
@miguelgrinberg sorry for the late reply, the error is on the client side. @saikrishb got this right I believe

**Comment by sambrightman** [comment]
I was having this issue and it was fixed by a recent engineio upgrade (3.2.3 -> 3.4.3 if I recall correctly).

**Comment by miguelgrinberg** [comment]
I'm going to close this, with the assumption that all the Python 2 problems have been addressed. If any new problems appear please open a new issue.
